File: python/pnl_tracker.py
import heapq

class PnLTracker:
    def __init__(self, market_prices):
        self.market_prices = market_prices
        self.positions = {}  # trader_id: {instrument: quantity}
        self.cash = {}       # trader_id: cash balance

    def process_fill(self, trader_id, instrument, quantity, price, side):
        if trader_id not in self.positions:
            self.positions[trader_id] = {}
            self.cash[trader_id] = 0

        if side == "buy":
            if instrument in self.positions[trader_id]:
                self.positions[trader_id][instrument] += quantity
            else:
                self.positions[trader_id][instrument] = quantity
            self.cash[trader_id] -= quantity * price
        elif side == "sell":
            # we are holding the instrument
            if instrument in self.positions[trader_id]:
                final_quantity = self.positions[trader_id][instrument] - quantity
                if final_quantity == 0:
                    del self.positions[trader_id][instrument]
                elif final_quantity > 0:
                    self.positions[trader_id][instrument] -= quantity
                self.cash[trader_id] += quantity * price

    def get_pnl(self, trader_id):
        if trader_id not in self.positions:
            return 0
        market_value = 0
        for instrument, quantity in self.positions[trader_id].items():
            market_value += quantity * self.market_prices[instrument]
        return market_value + self.cash[trader_id]

    def get_best_trader(self):
        best = None
        best_pnl = float('-inf')
        for trader_id in self.positions:
            pnl = self.get_pnl(trader_id)
            if pnl > best_pnl:
                best = trader_id
                best_pnl = pnl
        return best
    
    def get_top_k_traders(self, k):
        if k > len(self.positions):
            return "k too big"
        heap = []
        result = []
        for trader_id in self.positions:
            heapq.heappush_max(heap, (self.get_pnl(trader_id), trader_id))
        while len(result) < k:
            result.append(heapq.heappop_max(heap)[1])
        return result

    
tracker = PnLTracker({"AAPL": 150.0})

# trader sells entire position
tracker.process_fill("T1", "AAPL", 100, 150.0, "buy")
tracker.process_fill("T1", "AAPL", 100, 155.0, "sell")
print(tracker.cash["T1"])  # what does this print?

# all traders are losing money
tracker2 = PnLTracker({"AAPL": 100.0})
tracker2.process_fill("T1", "AAPL", 100, 150.0, "buy")
tracker2.process_fill("T2", "AAPL", 20000, 150.0, "buy")
tracker2.process_fill("T3", "AAPL", 400, 151.0, "buy")
tracker2.process_fill("T4", "AAPL", 500, 130.0, "buy")
print(tracker2.get_top_k_traders(2))
print(tracker2.get_best_trader())  # what does this print?

File: cpp/pnl_tracker.cpp
#include <iostream>
#include <queue>
#include <string>
#include <unordered_map>

enum class Side { Buy, Sell };

using hashmap = std::unordered_map<std::string, float>;

class PnLTracker {
	hashmap instruments;
	hashmap balances;
	std::unordered_map<std::string, std::unordered_map<std::string, float>> holdings;

public:
	PnLTracker(hashmap instruments) { this->instruments = instruments; }
	void process_fill(const std::string& trader, const std::string& instrument, float quantity, float price, Side side) {
		auto it = balances.find(trader);
		if (it == balances.end()) {
			std::cout << "Inserting new trader..." << std::endl;
			balances.insert({trader, 0});
			holdings.insert({trader, {}});
		}
		instruments[instrument] = price;
		float total = price * quantity;
		if (side == Side::Buy) {
			balances[trader] -= total;
			holdings[trader][instrument] += quantity;
		} else {
			balances[trader] += total;
			holdings[trader][instrument] -= quantity;
		}
	}
	float get_trader_pnl(const std::string& trader) {
		auto it = balances.find(trader);
		if (it == balances.end()) {
			std::cout << "Trader not registered" << std::endl;
			return 0;
		}
		float balance = balances[trader];
		float total_holdings = 0;
		for (auto it2 = holdings[trader].begin(); it2 != holdings[trader].end(); it2++) {
			std::string instrument = it2->first;
			float quantity = it2->second;
			total_holdings += (instruments[instrument] * quantity);
		}
		return balance + total_holdings;
	}
	std::string get_best_trader() {
		auto it = balances.begin();
		if (it == balances.end()) {
			return "No traders in system.";
		}
		float best_pnl = get_trader_pnl(it->first);
		std::string best_trader = it->first;
		float pnl;
		while (it != balances.end()) {
			pnl = get_trader_pnl(it->first);
			if (pnl > best_pnl) {
				best_pnl = pnl;
				best_trader = it->first;
			}
			it++;
		}
		return best_trader;
	}
	std::string* get_top_k_traders(size_t k) {
		std::priority_queue<std::pair<float, std::string>> pq;
		auto it = balances.begin();
		if (it == balances.end() || k > balances.size()) {
			return nullptr;
		}
		while (it != balances.end()) {
			pq.push(std::pair(get_trader_pnl(it->first), it->first));
			it++;
		}
		std::string* trader_names = new std::string[k];
		for (int i = 0; i < k; i++) {
			auto info = pq.top();
			pq.pop();
			trader_names[i] = info.second;
		}
		return trader_names;
	}
};

int main() {
	PnLTracker pnltracker({{"AAPL", 150.0}, {"MSFT", 100.5}});
	pnltracker.process_fill("John", "AAPL", 20, 145.0, Side::Buy);
	pnltracker.process_fill("John", "AAPL", 15, 144.0, Side::Sell);
	pnltracker.process_fill("Steve", "AAPL", 10, 144.0, Side::Buy);
	pnltracker.process_fill("Steve", "AAPL", 10, 146.0, Side::Sell);
	pnltracker.process_fill("Mark", "AAPL", 10, 146.0, Side::Buy);
	pnltracker.process_fill("Mark", "AAPL", 10, 150.0, Side::Sell);
	pnltracker.process_fill("John", "AAPL", 15, 150.0, Side::Buy);
	std::cout << pnltracker.get_trader_pnl("Steve") << std::endl;
	std::cout << pnltracker.get_best_trader() << std::endl;
	size_t k = 2;
	std::string* traders = pnltracker.get_top_k_traders(k);
	for (int i = 0; i < k; i++) {
		std::cout << traders[i] << std::endl;
	}
	delete[] traders;
	return 0;
}
